src/components/input/SelectNext.tsx

```diff
@@ -0,0 +1,182 @@
+import classnames from 'classnames';
+import type { ComponentChildren } from 'preact';
+import {
+  useCallback,
+  useContext,
+  useId,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from 'preact/hooks';
+
+import { useArrowKeyNavigation } from '../../hooks/use-arrow-key-navigation';
+import { useClickAway } from '../../hooks/use-click-away';
+import { useKeyPress } from '../../hooks/use-key-press';
+import { MenuCollapseIcon, MenuExpandIcon } from '../icons';
+import Button from './Button';
+import SelectContext from './SelectContext';
+
+export type SelectProps<T> = {
+  selected: T;
+  onChange: (newValue: T) => void;
+  label: ComponentChildren;
+  children: ComponentChildren;
+  disabled?: boolean;
+};
+
+function SelectMain<T>({
+  label,
+  selected,
+  onChange,
+  children,
+  disabled,
+}: SelectProps<T>) {
+  const [isDropdownOpen, setIsDropdownOpen] = useState(false);
+  const closeDropdown = useCallback(() => setIsDropdownOpen(false), []);
+  const wrapperRef = useRef<HTMLDivElement | null>(null);
+  const listboxId = useId();
+  const buttonRef = useRef<HTMLButtonElement | null>(null);
+  const buttonId = useId();
+  const [shouldDropUp, setShouldDropUp] = useState(false);
+
+  const selectValue = useCallback(
+    (newValue: unknown) => {
+      onChange(newValue as T);
+      closeDropdown();
+    },
+    [closeDropdown, onChange],
+  );
+
+  // When clicking away or pressing `Esc`, close the dropdown
+  useClickAway(wrapperRef, closeDropdown);
+  useKeyPress(['Escape'], closeDropdown);
+
+  // Vertical arrow key for options in the dropdown
+  useArrowKeyNavigation(wrapperRef, { horizontal: false });
+
+  // Focus button after closing dropdown
+  useLayoutEffect(() => {
+    if (!isDropdownOpen) {
+      buttonRef.current!.focus();
+    }
+  }, [isDropdownOpen]);
+
+  if (isDropdownOpen) {
+    const viewportHeight = window.innerHeight;
+    const { top: buttonDistanceToTop, height } =
+      buttonRef.current!.getBoundingClientRect();
```

Comment:
Reading the ref here happens to work because the `isDropdownOpen` state happens to always be false on the initial render, but it would be easier to break this assumption with future changes.